src/items/powerup_manager.cpp

```diff
@@ -205,142 +208,343 @@ void PowerupManager::LoadPowerup(PowerupType type, const XMLNode &node)
-        // over the powerups twice: first the single item version, then the
-        // multi-item version. The assignment to type takes care of this
-        m_powerups_for_position[pos_class].clear();
-        for(unsigned int i= POWERUP_FIRST; i<=2*POWERUP_LAST; i++)
-        {
-            PowerupType type =
-                (PowerupType) ((i<=POWERUP_LAST) ? i
-                                                 : i+POWERUP_FIRST);
-            unsigned int w =m_weights[pos_class][i-POWERUP_FIRST];
-            // The 'global' powerups (i.e. powerups that affect
-            // all karts, not only the ones close by) appear too
-            // frequently with larger number of karts. To reduce
-            // this effect their weight is reduced by the number
-            // of karts.
-            if(w!=0 && num_karts > 4 &&
-                 (type==POWERUP_PARACHUTE || type==POWERUP_SWITCH) )
-            {
-                w = w / (num_karts/4);
-                if(w==0) w=1;
-            }
-            for(unsigned int j=0; j<w; j++)
-                m_powerups_for_position[pos_class].push_back(type);
-        }   // for type in [POWERUP_FIRST, POWERUP_LAST]
+        m_position_to_class_inf.push_back(convertPositionToClass(num_karts, position, false));
+        m_position_to_class_sup.push_back(convertPositionToClass(num_karts, position, true));
+        m_position_to_class_cutoff.push_back(convertPositionToClassWeight(num_karts, position));
     }
 
+    // Then determine which items are available.
+    if (race_manager->isBattleMode())
+         updatePowerupClass(POSITION_BATTLE_MODE);
+    else if(race_manager->isSoccerMode())
+         updatePowerupClass(POSITION_SOCCER_MODE);
+    else if(race_manager->isTutorialMode())
+         updatePowerupClass(POSITION_TUTORIAL_MODE);
+    else
+    {
+        //TODO : replace this by a nice loop   
+        updatePowerupClass(POSITION_FIRST);
+        updatePowerupClass(POSITION_TOP33);
+        updatePowerupClass(POSITION_MID33);
+        updatePowerupClass(POSITION_END33);
+        updatePowerupClass(POSITION_LAST);
+    }
 }   // updateWeightsForRace
 
+// ----------------------------------------------------------------------------
+/** This function update the powerup array of a position class
+ *  \param pos_class The position class to update
+ */
+void PowerupManager::updatePowerupClass(PowerupManager::PositionClass pos_class)
+{
+
+    m_powerups_for_reference_pos[pos_class].clear();
+
+    // This loop actually goes over the powerups twice: first the single item version,
+    // then the multi-item version. The assignment to type takes care of this
+    for(unsigned int i= POWERUP_FIRST; i<=2*POWERUP_LAST; i++)
+    {
+        PowerupType type =
+            (PowerupType) ((i<=POWERUP_LAST) ? i
+                                             : i+POWERUP_FIRST);
+        unsigned int w =m_weights[pos_class][i-POWERUP_FIRST];
```

Comment:
Those lines are pretty weird. i + POWERUP_FIRST followed by i - POWERUP_FIRST, huh? Maybe this code is correct, but it's not obvious or easy to follow

Reply:
This is the same thing as the original loop, you can look it up on lines 286-296 of the file in master : https://github.com/supertuxkart/stk-code/blob/master/src/items/powerup_manager.cpp

The i+POWERUP_FIRST is in a ternary operator who is used to have the correct powerup type associated when the loop goes over the second part of the values to know how many triple-items there are.

The i-POWERUP_FIRST is used to convert the value from i to a correct index value, as POWERUP_FIRST, iirc, will be something like 1, while the first element of an array is at [0].